nibabel/spatialimages.py

```diff
@@ -461,8 +479,66 @@ def from_image(klass, img):
                      klass.header_class.from_header(img.header),
                      extra=img.extra.copy())
 
+    def slice_affine(self, idx):
+        """ Retrieve affine for current image, if sliced by a given index
+
+        Applies scaling if down-sampling is applied, and adjusts the intercept
+        to account for any cropping.
+
+        Parameters
+        ----------
+        idx : numpy-compatible slice index
+
+        Returns
+        -------
+        affine : (4,4) ndarray
+            Affine with updated scale and intercept
+        """
+        idx = canonical_slicers(idx, self.shape, check_inds=False)[:3]
+
+        # Transform:
+        # sx  0  0 tx
+        #  0 sy  0 ty
+        #  0  0 sz tz
+        #  0  0  0  1
+        transform = np.eye(4, dtype=int)
```

Comment:
By my algebra, to rescale, we do:

```Python
new_affine[:3, :3] = affine[:3, :3].dot([[sx, 0, 0], [0, sy, 0], [0, 0, sz]])
new_affine[:, [3]] = affine.dot([[tx], [ty], [tz], [1]])
```

Where `s*` are the steps, and `t*` are the starts.

Which I'm pretty sure is just equivalent to building this one transform matrix and doing `affine.dot(transform)`. If someone wants to check my math.